
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Simulation &#8212; CNV gene mapping</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Softwares and packages setup" href="setup.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">CNV gene mapping</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Overview
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="setup.html">
   Softwares and packages setup
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Simulation
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/minqiao/cnv-gene-mapping"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/minqiao/cnv-gene-mapping/issues/new?title=Issue%20on%20page%20%2Fworkflow.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/workflow.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.md</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview-of-the-workflow">
   Overview of the workflow
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#theory">
   Theory
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#study-design">
     Study design
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#parameters">
     Parameters
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#running-simulation">
   Running simulation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#workflow-overview">
     workflow overview
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#compiling-scripts">
     Compiling scripts
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#analysis-methods">
   Analysis methods
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#results">
   Results
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#histogram">
     Histogram
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#roc-curve">
     ROC curve
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#calibrated-pip-and-proportion-of-effect-genes">
     Calibrated PIP and proportion of effect genes
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Simulation</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview-of-the-workflow">
   Overview of the workflow
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#theory">
   Theory
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#study-design">
     Study design
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#parameters">
     Parameters
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#running-simulation">
   Running simulation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#workflow-overview">
     workflow overview
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#compiling-scripts">
     Compiling scripts
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#analysis-methods">
   Analysis methods
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#results">
   Results
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#histogram">
     Histogram
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#roc-curve">
     ROC curve
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#calibrated-pip-and-proportion-of-effect-genes">
     Calibrated PIP and proportion of effect genes
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="simulation">
<h1>Simulation<a class="headerlink" href="#simulation" title="Permalink to this headline">#</a></h1>
<section id="overview-of-the-workflow">
<h2>Overview of the workflow<a class="headerlink" href="#overview-of-the-workflow" title="Permalink to this headline">#</a></h2>
<p><img alt="" src="_images/workflow_diagram.png" /></p>
</section>
<section id="theory">
<h2>Theory<a class="headerlink" href="#theory" title="Permalink to this headline">#</a></h2>
<section id="study-design">
<h3>Study design<a class="headerlink" href="#study-design" title="Permalink to this headline">#</a></h3>
<p>We first use simulations to assess our methods. Since the simulated locations of CNVs need to match the sizes and frequencies of CNVs in real data, we start by partitioning real data of CNV events and genes from SCZ case-control study into CNV-gene genomic regions that harbor at least a pre-selected number of genes, for example, 30, in a region (figure or table). Next we generate a large number of simulated samples in parallel then concatenate them into whole genome X matrix. Subsequently, we sample risk genes by simulating effect size for each gene and sort simulated samples into cases and controls by logistic model. Finally, we partition the X matrix into “natural CNV-gene block” for gene-level fine-mapping and method evaluation. The detailed simulation process is as follows:</p>
<ul class="simple">
<li><p><strong>Step 1</strong>: Partition real data of CNV-gene into genomic regions to ensure each CNV-gene block contains at least 30 genes. The criteria for boundary gene of a particular region is that it must not overlap with any CNV events for all individuals. If the 30th gene overlaps with any CNV event for at least one individual, we check the next gene and see if it satisfies the criteria, till the nearest one does.</p></li>
<li><p><strong>Step 2</strong>: Simulate samples (<span class="math notranslate nohighlight">\(X\)</span> matrix) for deletion: sample each block from non-repetitive individuals and merge them together as a simulated individual. Repeat this process for <span class="math notranslate nohighlight">\(100,000\)</span> times to collect <span class="math notranslate nohighlight">\(100,000\)</span> individuals before simulating phenotype.</p></li>
<li><p><strong>Step 3</strong>: Simulate phenotype (<span class="math notranslate nohighlight">\(y\)</span> matrix): set prevalence as <span class="math notranslate nohighlight">\(0.05\)</span>, prevalence (<span class="math notranslate nohighlight">\(p\)</span>) <span class="math notranslate nohighlight">\(\approx \frac{e^{\beta_0}}{1-e^{\beta_0}}\)</span>, so <span class="math notranslate nohighlight">\(\beta_0 \approx \log \frac{p}{1-p}\)</span>. Odds ratio (OR) follows <span class="math notranslate nohighlight">\(e^{Normal(\mu,\sigma)}\)</span> or Gamma distribution, and <span class="math notranslate nohighlight">\(\beta_j = \text{log(OR)} \sim Normal(\mu,\sigma)\)</span>, then use Bernoulli (<span class="math notranslate nohighlight">\(\pi\)</span>) to decrease <span class="math notranslate nohighlight">\(95\)</span>% of <span class="math notranslate nohighlight">\(\beta_j\)</span>’s to 0.</p></li>
<li><p><strong>Step 4</strong>: Simulate <span class="math notranslate nohighlight">\(y\)</span>: <span class="math notranslate nohighlight">\(\text{logit}(y_i)=X_i\boldsymbol{\beta}+\beta_0\)</span>, <span class="math notranslate nohighlight">\(y_i=\frac{e^{x\boldsymbol{\beta}+\beta_0}}{1+e^{x\boldsymbol{\beta}+\beta_0}}\)</span> (<span class="math notranslate nohighlight">\(0&lt;y_i&lt;1\)</span>). Larger <span class="math notranslate nohighlight">\(y_i\)</span> indicates higher probability that it will be assigned as case. Then use Bernoulli (<span class="math notranslate nohighlight">\(y_i\)</span>) (will obtain 0 or 1) to classify <span class="math notranslate nohighlight">\(y_i\)</span> to either case (1) or control (0). Then select all 1’s (about <span class="math notranslate nohighlight">\(5\%\)</span>) as cases and randomly select equal number of 0’s as controls.</p></li>
<li><p><strong>Step 5</strong>: Partition the <span class="math notranslate nohighlight">\(X\)</span> matrix into “natural CNV-gene block” by the criteria in Step 1 without setting minimum number of genes in a block.</p></li>
</ul>
<p>We usually use 20 or 30 as the pre-selected minimum number of genes per genomic regions based on empirical experiences. Since it does not make obvious difference between 20 and 30, we choose 30 considering computational effort.</p>
</section>
<section id="parameters">
<h3>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline">#</a></h3>
<ol class="simple">
<li><p><span class="math notranslate nohighlight">\(\beta_j\)</span> ~ Normal (<span class="math notranslate nohighlight">\(0.77,0.84\)</span>). <span class="math notranslate nohighlight">\(0.77\ (\mu)\)</span> and <span class="math notranslate nohighlight">\(0.84\ (\sigma)\)</span> are calculated by <code class="docutils literal notranslate"><span class="pre">varbvs</span></code> over the whole genome.</p></li>
<li><p>penetrance = <span class="math notranslate nohighlight">\(0.05\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\pi = 0.043\)</span></p></li>
<li><p>Simulated sample size: <span class="math notranslate nohighlight">\(200,000\)</span></p></li>
</ol>
</section>
</section>
<section id="running-simulation">
<h2>Running simulation<a class="headerlink" href="#running-simulation" title="Permalink to this headline">#</a></h2>
<section id="workflow-overview">
<h3>workflow overview<a class="headerlink" href="#workflow-overview" title="Permalink to this headline">#</a></h3>
<p>Type command</p>
<p><code class="docutils literal notranslate"><span class="pre">sos</span> <span class="pre">run</span> <span class="pre">workflow/20190717_workflow.ipynb</span> <span class="pre">-h</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>usage: sos run workflow/20190717_workflow.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  genome_partition
  susie
  varbvs
  fisher
  mcmc
  mcmc_multichain
  sier
  hybrid
  get_hist
  varbvs_wg
  simulate

Global Workflow Options:
  --cwd output (as path)
  --genotype-file data/deletion.X.gz (as path)
  --phenotype-file data/deletion.y.gz (as path)
                        phenotype
  --name &#39;&#39;
  --blocks-file . (as path)
  --iteration 5000 (as int)
                        MCMC number of iterations
  --tune-prop 0.25 (as float)
                        MCMC ...
  --target-accept 0.98 (as float)
                        MCMC ...
  --n-chain 10 (as int)
                        MCMC ...
  --n-core 1 (as int)
                        MCMC ... For some reason on RCC cluster more than 1 cores will not work
                        (stuck)
  --[no-]reparameterize (default to False)
                        MCMC ...
  --prevalence 0.05 (as float)
                        alpha = log(p/1-p) is uniform lower bound: p = prevalence, when
                        prevalence = 0.05, alpha = -2.94 upper bound: p = case proportion, when
                        case = control, alpha = 0 MCMC ...
  --mcmc-seed 999 (as int)
                        MCMC ...
  --hyperparam-file . (as path)
                        Hyper-parameters for MCMC and for Single Effect Regression
  --L 10 (as int)
                        SuSiE number of effects
  --varbvs-wg-pip . (as path)
                        Whole genome PIPs obtained by varbvs using `varbvs_wg` pipeline, used for
                        hybrid pipeline
  --mcmc-walltime &#39;2.5h&#39;
                        cluster job configurations
  --sier-walltime 2h
  --job-size 80 (as int)

Sections
  genome_partition:
    Workflow Options:
      --input-file VAL (as str, required)
                        For simulation: get real deletion/duplication CNV data and its block
                        n_gene_in_block: get_hist: 1, simulate: 20~50, analyze: 1
      --output-files VAL VAL ... (as type, required)
                        output contain 3 files: 1) input data removing columns with all zeros, 2)
                        file containing block start and end matching index in 1), 3) block start
                        and end without reindex
      --n-gene-in-block VAL (as int, required)
                        minimum number of genes in a block for copy model set it to 1 to use
                        &quot;natural blocks&quot; from input data
      --col-index VAL (required)
                        col_index=None: no row names, col_index=0: use first column as row names
  susie_1, varbvs_1, fisher_1, mcmc_1, mcmc_multichain_1, sier_1, hybrid_1, get_hist_1:
  fisher_2:
  susie_2, varbvs_2, mcmc_2, mcmc_multichain_2, sier_2, hybrid_2:
  hybrid_3:
  mcmc_multichain_3:
  mcmc_3:
  sier_3:
    Workflow Options:
      --expected-effects -9 (as int)
  mcmc_4, mcmc_multichain_4, sier_4, hybrid_4:
  susie_3:
    Workflow Options:
      --estimate-prior-method simple
      --check-null-threshold 0.1 (as float)
  susie_4, varbvs_4:
  varbvs_3:
  varbvs_wg:
    Workflow Options:
      --maximum-prior-inclusion 0.0 (as float)
      --Rseed 999 (as int)
  get_hist_2:
  simulate_1:
    Workflow Options:
      --n-gene-in-block 30 (as int)
  simulate_2:
  simulate_3:
    Workflow Options:
      --shape 1.4 (as float)
      --scale 0.6 (as float)
      --beta-method normal
      --pi0 0.95 (as float)
      --seed 999 (as int)
  simulate_4:
    Workflow Options:
      --sample-size 100000 (as int)
      --n-batch 200 (as int)
  simulate_5:
  simulate_6:
</pre></div>
</div>
</section>
<section id="compiling-scripts">
<h3>Compiling scripts<a class="headerlink" href="#compiling-scripts" title="Permalink to this headline">#</a></h3>
<p>Run below shell script on your computer or remote server. Within our simulation framework, we</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you’re using remote host like slurm, please manually add corresponding command at frontmatter.</p>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">DATE</span><span class="o">=</span><span class="k">$(</span>date +%Y-%m-%d<span class="k">)</span>
<span class="nv">output</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$DATE</span><span class="s2">-OUTPUT&quot;</span>

<span class="c1">#simulate data</span>
sos run workflow/20190717_workflow.ipynb simulate <span class="se">\</span>
    --name simulation <span class="se">\</span>
    --genotype-file data/deletion.X.gz <span class="se">\</span>
    --cwd <span class="nv">$output</span> <span class="se">\</span>
    --sample-size <span class="m">200000</span> <span class="se">\</span>
    --n-batch <span class="m">200</span>
    --seed <span class="m">999999</span>
    
<span class="c1">#whole genome analysis using varbvs    </span>
sos run workflow/20190717_workflow.ipynb varbvs_wg <span class="se">\</span>
    --name varbvs --cwd <span class="nv">$output</span> <span class="se">\</span>
    --genotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.X.gz <span class="se">\</span>
    --phenotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.y.gz
    
<span class="c1">#Fisher&#39;s exact test per gene    </span>
sos run workflow/20190717_workflow.ipynb fisher <span class="se">\</span>
    --name fisher --cwd <span class="nv">$output</span> <span class="se">\</span>
    --genotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.X.gz <span class="se">\</span>
    --phenotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.y.gz
    
<span class="c1">#MCMC analysis</span>
sos run workflow/20190717_workflow.ipynb mcmc <span class="se">\</span>
    --name mcmc --cwd <span class="nv">$output</span> <span class="se">\</span>
    --genotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.X.gz <span class="se">\</span>
    --phenotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.y.gz <span class="se">\</span>
    --hyperparam-file <span class="nv">$output</span>/deletion.X_b30.simulation.varbvs.hyperparams                                                     ​

<span class="c1">#SuSiE analysis</span>
sos run workflow/20190717_workflow.ipynb susie <span class="se">\</span>
    --name susie --cwd <span class="nv">$output</span> <span class="se">\</span>
    --genotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.X.gz <span class="se">\</span>
    --phenotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.y.gz <span class="se">\</span>
    --hyperparam-file <span class="nv">$output</span>/deletion.X_b30.simulation.varbvs.hyperparams
    
<span class="c1">#Single Effect Regression analysis</span>
sos run workflow/20190717_workflow.ipynb sier <span class="se">\</span>
    --name sier --cwd <span class="nv">$output</span> <span class="se">\</span>
    --genotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.X.gz <span class="se">\</span>
    --phenotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.y.gz <span class="se">\</span>
    --hyperparam-file <span class="nv">$output</span>/deletion.X_b30.simulation.varbvs.hyperparams
    
<span class="c1">#varbvs analysis</span>
sos run workflow/20190717_workflow.ipynb varbvs <span class="se">\</span>
    --name varbvs --cwd <span class="nv">$output</span> <span class="se">\</span>
    --genotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.X.gz <span class="se">\</span>
    --phenotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.y.gz
   
<span class="c1">#Hybrid approach</span>
sos run workflow/20190717_workflow.ipynb hybrid <span class="se">\</span>
    --name hybrid --cwd <span class="nv">$output</span> <span class="se">\</span>
    --genotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.X.gz <span class="se">\</span>
    --phenotype-file <span class="nv">$output</span>/deletion.X_b30.simulation.y.gz <span class="se">\</span>
    --varbvs-wg-pip <span class="nv">$output</span>/deletion.X_b30.simulation.varbvs.pip <span class="se">\</span>
    --hyperparam-file <span class="nv">$output</span>/deletion.X_b30.simulation.varbvs.hyperparams

<span class="c1">#Get histogram</span>
sos run workflow/20190717_workflow.ipynb get_hist <span class="se">\</span>
    --name get_hist --cwd <span class="nv">$output</span>
</pre></div>
</div>
</section>
</section>
<section id="analysis-methods">
<h2>Analysis methods<a class="headerlink" href="#analysis-methods" title="Permalink to this headline">#</a></h2>
<ol class="simple">
<li><p>Divide the matrix of selected simulated samples (<span class="math notranslate nohighlight">\(23,856\)</span>) by genes (<span class="math notranslate nohighlight">\(23,343\)</span>) into blocks not interrupted by CNV events.</p></li>
<li><p>Apply the following methods in each CNV-gene block</p>
<ul class="simple">
<li><p><a class="reference external" href="https://mathworld.wolfram.com/FishersExactTest.html"><strong>Fisher’s exact test</strong></a>: generate 2 <span class="math notranslate nohighlight">\(\times\)</span> 2 table for each block and obtain p-value, four cells are # of cases with CNV event, # of cases without CNV event, # of controls with CNV event, # of controls without CNV event.</p></li>
<li><p><a class="reference external" href="https://www.biorxiv.org/content/10.1101/501114v2"><strong>SuSiE</strong></a>): Bayesian variable selection and genetic fine-mapping software package. It implements a simple new way to perform variable selection in multiple regression. It’s a particularly simple and intuitive fitting procedure – effectively a Bayesian modification of simple forward selection. The output of the fitting procedure is a number of “Credible Sets” (CSs), which are each designed to have high probability to contain a variable with non-zero effect, while at the same time being as small as possible.</p></li>
<li><p><a class="reference external" href="https://arxiv.org/pdf/1709.06597.pdf"><strong>varbvs</strong></a>: Large-Scale Bayesian Variable Selection Using Variational Methods.</p></li>
<li><p><strong>SIER</strong>: single implementation of the posterior computations for a “single effect” logistic regression with a normal prior on the unknown regression coefficients.</p>
<ul>
<li><p>It assumes one non-zero effect gene in each CNV-gene block. The PIPs summation in each block is 1.</p></li>
<li><p>A simpler solution: We can multiply the PIPs in each block by a specific “regional PIP” - the probability that the region has at least one causal effect.</p></li>
</ul>
</li>
<li><p><a class="reference external" href="https://peerj.com/articles/cs-55/"><strong>PyMC3</strong></a>: Probabilistic Programming in Python; Bayesian Modeling and Probabilistic Machine Learning with Theano. In our model, we use <a class="reference external" href="https://www.kaggle.com/melondonkey/bayesian-spike-and-slab-in-pymc3">Spike-and-Slab</a> prior to simulate logits.</p></li>
</ul>
</li>
</ol>
</section>
<section id="results">
<h2>Results<a class="headerlink" href="#results" title="Permalink to this headline">#</a></h2>
<section id="histogram">
<h3>Histogram<a class="headerlink" href="#histogram" title="Permalink to this headline">#</a></h3>
<p><img alt="" src="_images/hist.png" /></p>
</section>
<section id="roc-curve">
<h3>ROC curve<a class="headerlink" href="#roc-curve" title="Permalink to this headline">#</a></h3>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Before running the code, change the <code class="docutils literal notranslate"><span class="pre">path</span></code> to your local path.</p>
</div>
</aside>
<p>Start with building a PIP table (function <code class="docutils literal notranslate"><span class="pre">get_pip_table</span></code>) for different methods by using python with predefined number of blocks, here we use <span class="math notranslate nohighlight">\(30\)</span> blocks.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;/Users/bohan/cnv/2022-08-19-OUTPUT&quot;</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span><span class="o">,</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">Counter</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="n">n</span><span class="o">=</span><span class="mi">30</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">m1</span> <span class="o">=</span> <span class="s2">&quot;varbvs&quot;</span>
<span class="n">m2</span> <span class="o">=</span> <span class="s2">&quot;susie&quot;</span>
<span class="n">m3</span> <span class="o">=</span> <span class="s2">&quot;mcmc&quot;</span>
<span class="n">m4</span> <span class="o">=</span> <span class="s2">&quot;sier&quot;</span>
<span class="n">m5</span> <span class="o">=</span> <span class="s2">&quot;hybrid&quot;</span>
<span class="n">m6</span> <span class="o">=</span> <span class="s1">&#39;fisher&#39;</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_pip_table</span><span class="p">(</span><span class="n">n_blocks</span><span class="p">):</span>
    <span class="n">prefix</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;deletion.X_b</span><span class="si">{</span><span class="n">n_blocks</span><span class="si">}</span><span class="s2">.simulation.y&quot;</span>
    <span class="n">varbvs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">m1</span><span class="si">}</span><span class="s2">_pip.gz&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">usecols</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pip_</span><span class="si">{</span><span class="n">m1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span>
    <span class="n">susie</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">m2</span><span class="si">}</span><span class="s2">_pip.gz&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">usecols</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pip_</span><span class="si">{</span><span class="n">m2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span>
    <span class="n">mcmc</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">m3</span><span class="si">}</span><span class="s2">_pip.gz&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;pip_</span><span class="si">{</span><span class="n">m3</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span>
    <span class="n">sier</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">m4</span><span class="si">}</span><span class="s2">_pip.gz&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">usecols</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pip_</span><span class="si">{</span><span class="n">m4</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span>
    <span class="n">hybrid</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">m5</span><span class="si">}</span><span class="s2">_pip.gz&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">usecols</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pip_</span><span class="si">{</span><span class="n">m5</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span>
    <span class="n">fisher</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">prefix</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2">.X.cleaned.fisher.gz&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">beta_all</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">prefix</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2">.beta&quot;</span><span class="p">)</span>
    <span class="n">index</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">prefix</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2">.X.block_index_original.csv&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">])</span>
    <span class="n">betas</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">index</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">betas</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">beta_all</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]:(</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">)]))</span>
    <span class="n">pips</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">mcmc</span><span class="p">,</span> <span class="n">varbvs</span><span class="p">,</span> <span class="n">susie</span><span class="p">,</span> <span class="n">sier</span><span class="p">,</span> <span class="n">hybrid</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">fisher</span><span class="p">[[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;p&quot;</span><span class="p">]],</span> <span class="n">on</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span><span class="p">)</span>
    <span class="n">pips</span><span class="p">[</span><span class="s2">&quot;beta&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">betas</span>
    <span class="n">pips</span><span class="p">[</span><span class="s2">&quot;is_signal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pips</span><span class="p">[</span><span class="s2">&quot;beta&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">1</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pips</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pips</span> <span class="o">=</span> <span class="n">get_pip_table</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
<span class="n">pips</span> <span class="o">=</span> <span class="n">pips</span><span class="o">.</span><span class="n">set_index</span><span class="p">([[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">pips</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])]])</span>
<span class="n">pips</span><span class="p">[</span><span class="s1">&#39;pip_fisher&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">pips</span><span class="p">[</span><span class="s1">&#39;p&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Then, switch to R kernel and define a function <code class="docutils literal notranslate"><span class="pre">roc_data</span></code> which takes PIP values from the PIP table of each method as inputs and add confidence level 5% as cutoff, then outputs ROC data.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">roc_data</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="n">cols</span><span class="p">,</span> <span class="n">cutoff</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">0.05</span><span class="p">,</span> <span class="m">0.999</span><span class="p">),</span> <span class="n">connect_org</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">grid</span> <span class="o">=</span> <span class="m">100</span><span class="p">)</span> <span class="p">{</span>
    <span class="n">d1</span> <span class="o">=</span> <span class="n">dat</span><span class="p">[,</span><span class="n">cols</span><span class="p">]</span>
    <span class="nf">if </span><span class="p">(</span><span class="n">connect_org</span><span class="p">)</span> <span class="n">start</span> <span class="o">=</span> <span class="m">0</span>
    <span class="n">else</span> <span class="n">start</span> <span class="o">=</span> <span class="m">1</span>
    <span class="n">ttv</span> <span class="o">=</span> <span class="p">(</span><span class="n">start</span><span class="o">:</span><span class="n">grid</span><span class="p">)</span><span class="o">/</span><span class="n">grid</span>
    <span class="n">ttv</span> <span class="o">=</span> <span class="n">ttv</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="n">ttv</span><span class="o">&gt;=</span><span class="n">cutoff</span><span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="o">&amp;</span> <span class="n">ttv</span><span class="o">&lt;=</span><span class="n">cutoff</span><span class="p">[</span><span class="m">2</span><span class="p">])]</span>
    <span class="n">rst1</span> <span class="o">=</span> <span class="nf">t</span><span class="p">(</span><span class="nf">sapply</span><span class="p">(</span><span class="n">ttv</span><span class="p">,</span> <span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="nf">c</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">d1</span><span class="p">[,</span><span class="m">2</span><span class="p">][</span><span class="n">d1</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span><span class="o">&gt;=</span><span class="n">x</span><span class="p">]),</span> <span class="nf">length</span><span class="p">(</span><span class="n">d1</span><span class="p">[,</span><span class="m">2</span><span class="p">][</span><span class="n">d1</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span><span class="o">&gt;=</span><span class="n">x</span><span class="p">]),</span> <span class="nf">sum</span><span class="p">(</span><span class="n">d1</span><span class="p">[,</span><span class="m">2</span><span class="p">][</span><span class="n">d1</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span><span class="o">&gt;=</span><span class="n">x</span><span class="p">]</span><span class="o">==</span><span class="m">0</span><span class="p">))))</span>
    <span class="n">rst1</span> <span class="o">=</span> <span class="nf">cbind</span><span class="p">(</span><span class="n">rst1</span><span class="p">,</span> <span class="nf">sum</span><span class="p">(</span><span class="n">d1</span><span class="p">[,</span><span class="m">2</span><span class="p">]),</span> <span class="nf">sum</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">d1</span><span class="p">[,</span><span class="m">2</span><span class="p">]))</span>
    <span class="n">rst1</span> <span class="o">=</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="n">rst1</span><span class="p">)</span>
    <span class="nf">colnames</span><span class="p">(</span><span class="n">rst1</span><span class="p">)</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;true_positive&#39;</span><span class="p">,</span> <span class="s">&#39;total_positive&#39;</span><span class="p">,</span> <span class="s">&#39;false_positive&#39;</span><span class="p">,</span> <span class="s">&#39;total_signal&#39;</span><span class="p">,</span> <span class="s">&#39;total_null&#39;</span><span class="p">)</span>
    <span class="n">rst2</span> <span class="o">=</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">cbind</span><span class="p">(</span><span class="n">rst1</span><span class="o">$</span><span class="n">true_positive</span> <span class="o">/</span> <span class="n">rst1</span><span class="o">$</span><span class="n">total_positive</span><span class="p">,</span> <span class="n">rst1</span><span class="o">$</span><span class="n">true_positive</span> <span class="o">/</span> <span class="n">rst1</span><span class="o">$</span><span class="n">total_signal</span><span class="p">,</span>  <span class="n">ttv</span><span class="p">))</span>
    <span class="n">rst3</span> <span class="o">=</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">cbind</span><span class="p">(</span><span class="m">1</span> <span class="o">-</span> <span class="n">rst1</span><span class="o">$</span><span class="n">false_positive</span> <span class="o">/</span> <span class="n">rst1</span><span class="o">$</span><span class="n">total_null</span><span class="p">,</span> <span class="n">rst1</span><span class="o">$</span><span class="n">true_positive</span> <span class="o">/</span> <span class="n">rst1</span><span class="o">$</span><span class="n">total_signal</span><span class="p">,</span>  <span class="n">ttv</span><span class="p">))</span>
    <span class="nf">if </span><span class="p">(</span><span class="n">connect_org</span><span class="p">)</span> <span class="p">{</span>
        <span class="c1"># make a stair to origin</span>
        <span class="n">rst2</span> <span class="o">=</span> <span class="nf">rbind</span><span class="p">(</span><span class="n">rst2</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="nf">max</span><span class="p">(</span><span class="m">1</span> <span class="o">-</span> <span class="n">cutoff</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="n">rst2</span><span class="p">[</span><span class="nf">nrow</span><span class="p">(</span><span class="n">rst2</span><span class="p">),</span><span class="m">1</span><span class="p">]),</span> <span class="nf">max</span><span class="p">(</span><span class="n">rst2</span><span class="p">[</span><span class="nf">nrow</span><span class="p">(</span><span class="n">rst2</span><span class="p">),</span><span class="m">2</span><span class="p">]</span><span class="m">-0.01</span><span class="p">,</span> <span class="m">0</span><span class="p">),</span> <span class="n">rst2</span><span class="p">[</span><span class="nf">nrow</span><span class="p">(</span><span class="n">rst2</span><span class="p">),</span><span class="m">3</span><span class="p">]))</span>
        <span class="n">rst2</span> <span class="o">=</span> <span class="nf">rbind</span><span class="p">(</span><span class="n">rst2</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">1</span><span class="p">))</span>
        <span class="n">rst3</span> <span class="o">=</span> <span class="nf">rbind</span><span class="p">(</span><span class="n">rst3</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">1</span><span class="p">))</span>
    <span class="p">}</span>
    <span class="nf">colnames</span><span class="p">(</span><span class="n">rst2</span><span class="p">)</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;Precision&#39;</span><span class="p">,</span> <span class="s">&#39;Recall&#39;</span><span class="p">,</span> <span class="s">&#39;Threshold&#39;</span><span class="p">)</span>
    <span class="nf">colnames</span><span class="p">(</span><span class="n">rst3</span><span class="p">)</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;TN&#39;</span><span class="p">,</span> <span class="s">&#39;TP&#39;</span><span class="p">,</span> <span class="s">&#39;Threshold&#39;</span><span class="p">)</span>
    <span class="nf">return</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">counts</span> <span class="o">=</span> <span class="n">rst1</span><span class="p">,</span> <span class="n">pr</span> <span class="o">=</span> <span class="n">rst2</span><span class="p">,</span> <span class="n">roc</span> <span class="o">=</span> <span class="n">rst3</span><span class="p">))</span>
<span class="p">}</span>
</pre></div>
</div>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span>%<span class="n">get</span> <span class="n">pips</span> <span class="n">m1</span> <span class="n">m2</span> <span class="n">m3</span> <span class="n">m4</span> <span class="n">m5</span> <span class="n">m6</span>
<span class="n">res</span> <span class="o">=</span> <span class="nf">list</span><span class="p">()</span>
<span class="nf">for </span><span class="p">(</span><span class="n">m</span> <span class="n">in</span> <span class="nf">c</span><span class="p">(</span><span class="n">m1</span><span class="p">,</span><span class="n">m2</span><span class="p">,</span><span class="n">m3</span><span class="p">,</span><span class="n">m4</span><span class="p">,</span><span class="n">m5</span><span class="p">,</span><span class="n">m6</span><span class="p">))</span> <span class="n">res</span><span class="p">[[</span><span class="n">m</span><span class="p">]]</span> <span class="o">=</span> <span class="nf">roc_data</span><span class="p">(</span><span class="n">pips</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#39;pip_&#39;</span><span class="p">,</span><span class="n">m</span><span class="p">),</span> <span class="s">&#39;is_signal&#39;</span><span class="p">),</span> <span class="n">cutoff</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">1</span><span class="p">),</span> <span class="n">connect_org</span> <span class="o">=</span> <span class="bp">T</span><span class="p">,</span> <span class="n">grid</span><span class="o">=</span><span class="m">200</span><span class="p">)</span>
</pre></div>
</div>
<p>Lastly, switch back to SoS kernel and <code class="docutils literal notranslate"><span class="pre">%get</span></code> ROC data from R kernel and plot them.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">get</span> <span class="n">res</span> <span class="o">--</span><span class="kn">from</span> <span class="nn">R</span>

<span class="kn">import</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib.font_manager</span> <span class="k">as</span> <span class="nn">font_manager</span>
<span class="n">font_prop</span> <span class="o">=</span> <span class="n">font_manager</span><span class="o">.</span><span class="n">FontProperties</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">24</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="k">for</span> <span class="n">m</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="n">m1</span><span class="p">,</span><span class="n">m2</span><span class="p">,</span><span class="n">m3</span><span class="p">,</span><span class="n">m4</span><span class="p">,</span><span class="n">m5</span><span class="p">,</span><span class="n">m6</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;C0&#39;</span><span class="p">,</span><span class="s1">&#39;C8&#39;</span><span class="p">,</span><span class="s1">&#39;C3&#39;</span><span class="p">,</span><span class="s1">&#39;C4&#39;</span><span class="p">,</span><span class="s1">&#39;C2&#39;</span><span class="p">,</span><span class="s1">&#39;C7&#39;</span><span class="p">]):</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">res</span><span class="p">[</span><span class="n">m</span><span class="p">][</span><span class="s1">&#39;roc&#39;</span><span class="p">][</span><span class="s1">&#39;TN&#39;</span><span class="p">],</span> <span class="n">res</span><span class="p">[</span><span class="n">m</span><span class="p">][</span><span class="s1">&#39;roc&#39;</span><span class="p">][</span><span class="s1">&#39;TP&#39;</span><span class="p">],</span> <span class="n">c</span> <span class="o">=</span> <span class="n">c</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="n">m</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span> <span class="o">=</span> <span class="s1">&#39;lower right&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">15</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Power&quot;</span><span class="p">,</span> <span class="n">fontproperties</span><span class="o">=</span><span class="n">font_prop</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;False positive&quot;</span><span class="p">,</span> <span class="n">fontproperties</span><span class="o">=</span><span class="n">font_prop</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;ROC curve&#39;</span><span class="p">,</span> <span class="n">fontproperties</span><span class="o">=</span><span class="n">font_prop</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/ROC_curve.pdf&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="text-align:center head"><p><span class="math notranslate nohighlight">\(n=30,000\)</span></p></th>
<th class="text-align:center head"><p><span class="math notranslate nohighlight">\(n=80,000\)</span></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-align:center"><p><img alt="" src="_images/ROC_4simu_0802.png" /></p></td>
<td class="text-align:center"><p><img alt="" src="_images/ROC_large_sample_0802_12.png" /></p></td>
</tr>
</tbody>
</table>
</section>
<section id="calibrated-pip-and-proportion-of-effect-genes">
<h3>Calibrated PIP and proportion of effect genes<a class="headerlink" href="#calibrated-pip-and-proportion-of-effect-genes" title="Permalink to this headline">#</a></h3>
<p>Selecting the column of indicator and columns of all methods except Fisher’s exact test from PIP table.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">calib</span> <span class="o">=</span> <span class="n">pips</span><span class="p">[[</span><span class="sa">f</span><span class="s2">&quot;pip_</span><span class="si">{</span><span class="n">m1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;pip_</span><span class="si">{</span><span class="n">m2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;pip_</span><span class="si">{</span><span class="n">m3</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;pip_</span><span class="si">{</span><span class="n">m4</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;pip_</span><span class="si">{</span><span class="n">m5</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;is_signal&quot;</span><span class="p">]]</span>
</pre></div>
</div>
<p>Set bin size by sample size, we choose <span class="math notranslate nohighlight">\(10\)</span> as we have 200,000 samples.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span>%<span class="n">get</span> <span class="n">calib</span>

<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">cowplot</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>

<span class="n">bin_size</span> <span class="o">=</span> <span class="m">10</span>
<span class="n">bins</span> <span class="o">=</span> <span class="nf">cbind</span><span class="p">(</span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="n">bin_size</span><span class="p">)</span><span class="o">/</span><span class="n">bin_size</span><span class="m">-1</span><span class="o">/</span><span class="n">bin_size</span><span class="p">,</span> <span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="n">bin_size</span><span class="p">)</span><span class="o">/</span><span class="n">bin_size</span><span class="p">)</span>
</pre></div>
</div>
<p>In R kernel, pick all susceptible genes for each method by looping all bin sizes and calculate proportion.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">pip_cali</span> <span class="o">=</span> <span class="nf">list</span><span class="p">()</span>
<span class="nf">for </span><span class="p">(</span><span class="n">name</span> <span class="n">in</span> <span class="nf">rev</span><span class="p">(</span><span class="nf">colnames</span><span class="p">(</span><span class="n">calib</span><span class="p">))[</span><span class="m">-1</span><span class="p">])</span> <span class="p">{</span>
    <span class="nf">for </span><span class="p">(</span><span class="n">i</span> <span class="n">in</span> <span class="m">1</span><span class="o">:</span><span class="nf">nrow</span><span class="p">(</span><span class="n">bins</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">calib</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="n">calib</span><span class="p">[[</span><span class="n">name</span><span class="p">]]</span> <span class="o">&gt;</span> <span class="n">bins</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="m">1</span><span class="p">]</span> <span class="o">&amp;</span> <span class="n">calib</span><span class="p">[[</span><span class="n">name</span><span class="p">]]</span> <span class="o">&lt;</span> <span class="n">bins</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="m">2</span><span class="p">]),</span> <span class="p">]</span>
        <span class="nf">if </span><span class="p">(</span><span class="nf">is.null</span><span class="p">(</span><span class="n">pip_cali</span><span class="p">[[</span><span class="n">name</span><span class="p">]]))</span> <span class="n">pip_cali</span><span class="p">[[</span><span class="n">name</span><span class="p">]]</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">tmp</span><span class="p">[[</span><span class="n">name</span><span class="p">]]),</span> <span class="nf">sum</span><span class="p">(</span><span class="n">tmp</span><span class="o">$</span><span class="n">is_signal</span><span class="p">),</span> <span class="nf">length</span><span class="p">(</span><span class="n">tmp</span><span class="o">$</span><span class="n">is_signal</span><span class="p">))</span>
        <span class="n">else</span> <span class="n">pip_cali</span><span class="p">[[</span><span class="n">name</span><span class="p">]]</span> <span class="o">=</span> <span class="nf">rbind</span><span class="p">(</span><span class="n">pip_cali</span><span class="p">[[</span><span class="n">name</span><span class="p">]],</span> <span class="nf">c</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">tmp</span><span class="p">[[</span><span class="n">name</span><span class="p">]]),</span> <span class="nf">sum</span><span class="p">(</span><span class="n">tmp</span><span class="o">$</span><span class="n">is_signal</span><span class="p">),</span> <span class="nf">length</span><span class="p">(</span><span class="n">tmp</span><span class="o">$</span><span class="n">is_signal</span><span class="p">)))</span>
    <span class="p">}</span>
    <span class="n">pip_cali</span><span class="p">[[</span><span class="n">name</span><span class="p">]][</span><span class="nf">which</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">pip_cali</span><span class="p">[[</span><span class="n">name</span><span class="p">]]))]</span> <span class="o">=</span> <span class="m">0</span>
<span class="p">}</span>

<span class="n">get_cali</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">alist</span><span class="p">,</span> <span class="n">col</span><span class="p">)</span> <span class="p">{</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">alist</span><span class="p">[[</span><span class="n">col</span><span class="p">]]</span>
    <span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">alist</span><span class="p">[[</span><span class="n">col</span><span class="p">]]))</span> <span class="n">res</span> <span class="o">=</span> <span class="n">res</span> <span class="o">+</span> <span class="n">alist</span><span class="p">[[</span><span class="n">col</span><span class="p">]]</span>
    <span class="n">res</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">)]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">)]</span> <span class="o">/</span> <span class="n">res</span><span class="p">[,</span><span class="m">3</span><span class="p">]</span>
    <span class="nf">return</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="m">-1</span><span class="p">,])</span>
<span class="p">}</span>
</pre></div>
</div>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">dat</span> <span class="o">=</span> <span class="nf">list</span><span class="p">(</span><span class="s">&quot;varbvs&quot;</span> <span class="o">=</span> <span class="nf">get_cali</span><span class="p">(</span><span class="n">pip_cali</span><span class="p">,</span> <span class="s">&#39;pip_varbvs&#39;</span><span class="p">),</span>
           <span class="s">&quot;susie&quot;</span> <span class="o">=</span> <span class="nf">get_cali</span><span class="p">(</span><span class="n">pip_cali</span><span class="p">,</span> <span class="s">&#39;pip_susie&#39;</span><span class="p">),</span>
           <span class="s">&quot;pymc3&quot;</span> <span class="o">=</span> <span class="nf">get_cali</span><span class="p">(</span><span class="n">pip_cali</span><span class="p">,</span> <span class="s">&#39;pip_mcmc&#39;</span><span class="p">),</span> 
           <span class="s">&quot;sier&quot;</span> <span class="o">=</span> <span class="nf">get_cali</span><span class="p">(</span><span class="n">pip_cali</span><span class="p">,</span> <span class="s">&#39;pip_sier&#39;</span><span class="p">),</span>
           <span class="s">&quot;hybrid&quot;</span> <span class="o">=</span> <span class="nf">get_cali</span><span class="p">(</span><span class="n">pip_cali</span><span class="p">,</span> <span class="s">&#39;pip_hybrid&#39;</span><span class="p">))</span>

<span class="n">dot_plot</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">dataframe</span><span class="p">)</span> <span class="p">{</span>
  <span class="nf">ggplot</span><span class="p">(</span><span class="n">dataframe</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">mean_pip</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">observed_freq</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_errorbar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">ymin</span><span class="o">=</span><span class="n">observed_freq</span><span class="o">-</span><span class="n">se</span><span class="p">,</span> <span class="n">ymax</span><span class="o">=</span><span class="n">observed_freq</span><span class="o">+</span><span class="n">se</span><span class="p">),</span> <span class="n">colour</span><span class="o">=</span><span class="s">&quot;gray&quot;</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="m">0.2</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">.</span><span class="m">01</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">geom_point</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="m">1.5</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="m">21</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="s">&quot;#002b36&quot;</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;Mean PIP&quot;</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">ylab</span><span class="p">(</span><span class="s">&quot;Observed frequency&quot;</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">coord_cartesian</span><span class="p">(</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">),</span> <span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_abline</span><span class="p">(</span><span class="n">slope</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="n">intercept</span><span class="o">=</span><span class="m">0</span><span class="p">,</span><span class="n">colour</span><span class="o">=</span><span class="s">&#39;red&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="m">0.2</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">ggtitle</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">expand_limits</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="m">0</span><span class="p">)</span> <span class="o">+</span> 
    <span class="nf">theme_cowplot</span><span class="p">()</span>
<span class="p">}</span>
</pre></div>
</div>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>As before, change the <code class="docutils literal notranslate"><span class="pre">path</span></code> to your local path before plotting.</p>
</div>
</aside>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">path</span> <span class="o">=</span> <span class="s">&quot;/Users/bohan/cnv/2022-08-19-OUTPUT/&quot;</span>
<span class="n">idx</span> <span class="o">=</span> <span class="m">0</span>
<span class="nf">for </span><span class="p">(</span><span class="n">name</span> <span class="n">in</span> <span class="nf">names</span><span class="p">(</span><span class="n">dat</span><span class="p">))</span> <span class="p">{</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">idx</span> <span class="o">+</span> <span class="m">1</span>
    <span class="n">dat</span><span class="p">[[</span><span class="n">name</span><span class="p">]][,</span><span class="m">3</span><span class="p">]</span> <span class="o">=</span> <span class="nf">sqrt</span><span class="p">(</span><span class="n">dat</span><span class="p">[[</span><span class="n">name</span><span class="p">]][,</span><span class="m">2</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="m">1</span> <span class="o">-</span> <span class="n">dat</span><span class="p">[[</span><span class="n">name</span><span class="p">]][,</span><span class="m">2</span><span class="p">])</span> <span class="o">/</span> <span class="n">dat</span><span class="p">[[</span><span class="n">name</span><span class="p">]][,</span><span class="m">3</span><span class="p">])</span> <span class="o">*</span> <span class="m">2</span>
    <span class="n">dat</span><span class="p">[[</span><span class="n">name</span><span class="p">]]</span> <span class="o">=</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="n">dat</span><span class="p">[[</span><span class="n">name</span><span class="p">]])</span>
    <span class="nf">colnames</span><span class="p">(</span><span class="n">dat</span><span class="p">[[</span><span class="n">name</span><span class="p">]])</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;mean_pip&quot;</span><span class="p">,</span> <span class="s">&quot;observed_freq&quot;</span><span class="p">,</span> <span class="s">&quot;se&quot;</span><span class="p">)</span>
    <span class="nf">pdf</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s">&#39;_&#39;</span> <span class="p">,</span> <span class="n">bin_size</span><span class="p">,</span> <span class="s">&#39;_&#39;</span><span class="p">,</span> <span class="s">&#39;.pdf&#39;</span><span class="p">),</span> <span class="n">width</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">pointsize</span><span class="o">=</span><span class="m">16</span><span class="p">)</span>
    <span class="nf">print</span><span class="p">(</span><span class="nf">dot_plot</span><span class="p">(</span><span class="n">dat</span><span class="p">[[</span><span class="n">name</span><span class="p">]]))</span>
    <span class="nf">dev.off</span><span class="p">()</span>
    <span class="nf">system</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;convert -flatten -density 120 &quot;</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s">&#39;_&#39;</span> <span class="p">,</span> <span class="n">bin_size</span><span class="p">,</span> <span class="s">&#39;_&#39;</span><span class="p">,</span> <span class="s">&#39;.pdf&#39;</span><span class="p">,</span> <span class="s">&quot; &quot;</span><span class="p">,</span> 
                  <span class="n">path</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s">&#39;_&#39;</span> <span class="p">,</span> <span class="n">bin_size</span><span class="p">,</span> <span class="s">&#39;_&#39;</span><span class="p">,</span> <span class="s">&#39;.png&#39;</span><span class="p">))</span>
    <span class="p">}</span>
</pre></div>
</div>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p><img alt="alt" src="_images/hybrid_11_simu123_0709.png" /></p></th>
<th class="head"><p><img alt="alt" src="_images/MCMC_11_simu123_0709.png" /></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><img alt="alt" src="_images/SuSiE_12_simu123_0709.png" /></p></td>
<td><p><img alt="alt" src="_images/SIER_11_simu123_0709.png" /></p></td>
</tr>
</tbody>
</table>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="setup.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Softwares and packages setup</p>
        </div>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Gao Wang, Min Qiao, Bohan Jiang<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>