{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ba163511",
   "metadata": {},
   "source": [
    "# Workflow of simulation process\n",
    "\n",
    "## Logistics of the workflow\n",
    "![](cnv-figures/workflow_diagram.png)\n",
    "\n",
    "## Overview of the workflow\n",
    "Type command\n",
    "\n",
    "`sos run workflow/20190717_workflow.ipynb -h`\n",
    "\n",
    "```\n",
    "usage: sos run workflow/20190717_workflow.ipynb\n",
    "               [workflow_name | -t targets] [options] [workflow_options]\n",
    "  workflow_name:        Single or combined workflows defined in this script\n",
    "  targets:              One or more targets to generate\n",
    "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\n",
    "  workflow_options:     Double-hyphen workflow-specific parameters\n",
    "\n",
    "Workflows:\n",
    "  genome_partition\n",
    "  susie\n",
    "  varbvs\n",
    "  fisher\n",
    "  mcmc\n",
    "  mcmc_multichain\n",
    "  sier\n",
    "  hybrid\n",
    "  get_hist\n",
    "  varbvs_wg\n",
    "  simulate\n",
    "\n",
    "Global Workflow Options:\n",
    "  --cwd output (as path)\n",
    "  --genotype-file data/deletion.X.gz (as path)\n",
    "  --phenotype-file data/deletion.y.gz (as path)\n",
    "                        phenotype\n",
    "  --name ''\n",
    "  --blocks-file . (as path)\n",
    "  --iteration 5000 (as int)\n",
    "                        MCMC number of iterations\n",
    "  --tune-prop 0.25 (as float)\n",
    "                        MCMC ...\n",
    "  --target-accept 0.98 (as float)\n",
    "                        MCMC ...\n",
    "  --n-chain 10 (as int)\n",
    "                        MCMC ...\n",
    "  --n-core 1 (as int)\n",
    "                        MCMC ... For some reason on RCC cluster more than 1 cores will not work\n",
    "                        (stuck)\n",
    "  --[no-]reparameterize (default to False)\n",
    "                        MCMC ...\n",
    "  --prevalence 0.05 (as float)\n",
    "                        alpha = log(p/1-p) is uniform lower bound: p = prevalence, when\n",
    "                        prevalence = 0.05, alpha = -2.94 upper bound: p = case proportion, when\n",
    "                        case = control, alpha = 0 MCMC ...\n",
    "  --mcmc-seed 999 (as int)\n",
    "                        MCMC ...\n",
    "  --hyperparam-file . (as path)\n",
    "                        Hyper-parameters for MCMC and for Single Effect Regression\n",
    "  --L 10 (as int)\n",
    "                        SuSiE number of effects\n",
    "  --varbvs-wg-pip . (as path)\n",
    "                        Whole genome PIPs obtained by varbvs using `varbvs_wg` pipeline, used for\n",
    "                        hybrid pipeline\n",
    "  --mcmc-walltime '2.5h'\n",
    "                        cluster job configurations\n",
    "  --sier-walltime 2h\n",
    "  --job-size 80 (as int)\n",
    "\n",
    "Sections\n",
    "  genome_partition:\n",
    "    Workflow Options:\n",
    "      --input-file VAL (as str, required)\n",
    "                        For simulation: get real deletion/duplication CNV data and its block\n",
    "                        n_gene_in_block: get_hist: 1, simulate: 20~50, analyze: 1\n",
    "      --output-files VAL VAL ... (as type, required)\n",
    "                        output contain 3 files: 1) input data removing columns with all zeros, 2)\n",
    "                        file containing block start and end matching index in 1), 3) block start\n",
    "                        and end without reindex\n",
    "      --n-gene-in-block VAL (as int, required)\n",
    "                        minimum number of genes in a block for copy model set it to 1 to use\n",
    "                        \"natural blocks\" from input data\n",
    "      --col-index VAL (required)\n",
    "                        col_index=None: no row names, col_index=0: use first column as row names\n",
    "  susie_1, varbvs_1, fisher_1, mcmc_1, mcmc_multichain_1, sier_1, hybrid_1, get_hist_1:\n",
    "  fisher_2:\n",
    "  susie_2, varbvs_2, mcmc_2, mcmc_multichain_2, sier_2, hybrid_2:\n",
    "  hybrid_3:\n",
    "  mcmc_multichain_3:\n",
    "  mcmc_3:\n",
    "  sier_3:\n",
    "    Workflow Options:\n",
    "      --expected-effects -9 (as int)\n",
    "  mcmc_4, mcmc_multichain_4, sier_4, hybrid_4:\n",
    "  susie_3:\n",
    "    Workflow Options:\n",
    "      --estimate-prior-method simple\n",
    "      --check-null-threshold 0.1 (as float)\n",
    "  susie_4, varbvs_4:\n",
    "  varbvs_3:\n",
    "  varbvs_wg:\n",
    "    Workflow Options:\n",
    "      --maximum-prior-inclusion 0.0 (as float)\n",
    "      --Rseed 999 (as int)\n",
    "  get_hist_2:\n",
    "  simulate_1:\n",
    "    Workflow Options:\n",
    "      --n-gene-in-block 30 (as int)\n",
    "  simulate_2:\n",
    "  simulate_3:\n",
    "    Workflow Options:\n",
    "      --shape 1.4 (as float)\n",
    "      --scale 0.6 (as float)\n",
    "      --beta-method normal\n",
    "      --pi0 0.95 (as float)\n",
    "      --seed 999 (as int)\n",
    "  simulate_4:\n",
    "    Workflow Options:\n",
    "      --sample-size 100000 (as int)\n",
    "      --n-batch 200 (as int)\n",
    "  simulate_5:\n",
    "  simulate_6:\n",
    "```\n",
    "\n",
    "## Command\n",
    "\n",
    "### Obtain simulation results, run shell script below\n",
    "```\n",
    "DATE=$(date +%Y-%m-%d)\n",
    "output=\"$DATE-OUTPUT\"\n",
    "\n",
    "#simulate data\n",
    "sos run workflow/20190717_workflow.ipynb simulate \\\n",
    "    --name simulation \\\n",
    "    --genotype-file data/deletion.X.gz \\\n",
    "    --cwd $output \\\n",
    "    --sample-size 200000 \\\n",
    "    --n-batch 200\n",
    "    --seed 999999\n",
    "    \n",
    "#whole genome analysis using varbvs    \n",
    "sos run workflow/20190717_workflow.ipynb varbvs_wg \\\n",
    "    --name varbvs --cwd $output \\\n",
    "    --genotype-file $output/deletion.X_b30.simulation.X.gz \\\n",
    "    --phenotype-file $output/deletion.X_b30.simulation.y.gz\n",
    "    \n",
    "#Fisher's exact test per gene    \n",
    "sos run workflow/20190717_workflow.ipynb fisher \\\n",
    "    --name fisher --cwd $output \\\n",
    "    --genotype-file $output/deletion.X_b30.simulation.X.gz \\\n",
    "    --phenotype-file $output/deletion.X_b30.simulation.y.gz\n",
    "    \n",
    "#MCMC analysis\n",
    "sos run workflow/20190717_workflow.ipynb mcmc \\\n",
    "    --name mcmc --cwd $output \\\n",
    "    --genotype-file $output/deletion.X_b30.simulation.X.gz \\\n",
    "    --phenotype-file $output/deletion.X_b30.simulation.y.gz \\\n",
    "    --hyperparam-file $output/deletion.X_b30.simulation.varbvs.hyperparams                                                     â€‹\n",
    "\n",
    "#SuSiE analysis\n",
    "sos run workflow/20190717_workflow.ipynb susie \\\n",
    "    --name susie --cwd $output \\\n",
    "    --genotype-file $output/deletion.X_b30.simulation.X.gz \\\n",
    "    --phenotype-file $output/deletion.X_b30.simulation.y.gz \\\n",
    "    --hyperparam-file $output/deletion.X_b30.simulation.varbvs.hyperparams\n",
    "    \n",
    "#Single Effect Regression analysis\n",
    "sos run workflow/20190717_workflow.ipynb sier \\\n",
    "    --name sier --cwd $output \\\n",
    "    --genotype-file $output/deletion.X_b30.simulation.X.gz \\\n",
    "    --phenotype-file $output/deletion.X_b30.simulation.y.gz \\\n",
    "    --hyperparam-file $output/deletion.X_b30.simulation.varbvs.hyperparams\n",
    "    \n",
    "#varbvs analysis\n",
    "sos run workflow/20190717_workflow.ipynb varbvs \\\n",
    "    --name varbvs --cwd $output \\\n",
    "    --genotype-file $output/deletion.X_b30.simulation.X.gz \\\n",
    "    --phenotype-file $output/deletion.X_b30.simulation.y.gz\n",
    "   \n",
    "#Hybrid approach\n",
    "sos run workflow/20190717_workflow.ipynb hybrid \\\n",
    "    --name hybrid --cwd $output \\\n",
    "    --genotype-file $output/deletion.X_b30.simulation.X.gz \\\n",
    "    --phenotype-file $output/deletion.X_b30.simulation.y.gz \\\n",
    "    --varbvs-wg-pip $output/deletion.X_b30.simulation.varbvs.pip \\\n",
    "    --hyperparam-file $output/deletion.X_b30.simulation.varbvs.hyperparams\n",
    "\n",
    "#Get histogram\n",
    "sos run workflow/20190717_workflow.ipynb get_hist \\\n",
    "    --name get_hist --cwd $output\n",
    "```\n",
    "\n",
    "## Results\n",
    "\n",
    "### Histogram\n",
    "![](cnv-figures/hist.png)\n",
    "\n",
    "### ROC curve"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bbd5d11",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"/Users/bohan/cnv/2022-08-19-OUTPUT\"\n",
    "import pandas as pd, numpy as np\n",
    "import os\n",
    "from collections import Counter\n",
    "cwd = os.path.expanduser(path)\n",
    "n = 30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "655a57d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "m1 = \"varbvs\"\n",
    "m2 = \"susie\"\n",
    "m3 = \"mcmc\"\n",
    "m4 = \"sier\"\n",
    "m5 = \"hybrid\"\n",
    "m6 = 'fisher'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5df5245",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pip_table():\n",
    "    prefix = \"deletion.X_b30.simulation.y\"\n",
    "    varbvs = pd.read_csv(f\"{cwd}/{prefix}.{m1}_pip.gz\", sep = \"\\t\", header = None, usecols = [1], names = [f\"pip_{m1}\"])\n",
    "    susie = pd.read_csv(f\"{cwd}/{prefix}.{m2}_pip.gz\", sep = \"\\t\", header = None, usecols = [1], names = [f\"pip_{m2}\"])\n",
    "    mcmc = pd.read_csv(f\"{cwd}/{prefix}.{m3}_pip.gz\", sep = \"\\t\", header = None, names = [\"gene\", f\"pip_{m3}\"])\n",
    "    sier = pd.read_csv(f\"{cwd}/{prefix}.{m4}_pip.gz\", sep = \"\\t\", header = None, usecols = [1], names = [f\"pip_{m4}\"])\n",
    "    hybrid = pd.read_csv(f\"{cwd}/{prefix}.{m5}_pip.gz\", sep = \"\\t\", header = None, usecols = [1], names = [f\"pip_{m5}\"])\n",
    "    fisher = pd.read_csv(f\"{cwd}/{prefix[0:-2]}.X.cleaned.fisher.gz\", sep = \"\\t\", header = 0)\n",
    "    beta_all = np.loadtxt(f\"{cwd}/{prefix[0:-2]}.beta\")\n",
    "    index = pd.read_csv(f\"{cwd}/{prefix[0:-2]}.X.block_index_original.csv\", sep = \"\\t\", header = None, names = [\"start\", \"end\"])\n",
    "    betas = list()\n",
    "    for i, item in index.iterrows():\n",
    "        betas.extend(list(beta_all[item[0]:(item[1]+1)]))\n",
    "    pips = pd.concat([mcmc, varbvs, susie, sier, hybrid], axis = 1).merge(fisher[[\"gene\", \"p\"]], on = \"gene\")\n",
    "    pips[\"beta\"] = betas\n",
    "    pips[\"is_signal\"] = pips[\"beta\"].apply(lambda x: 1 if x != 0 else 0)\n",
    "    return pips"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "101b115b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pips = get_pip_table()\n",
    "pips = pips.set_index([[i for i in range(pips.shape[0])]])\n",
    "pips['pip_fisher'] = 1 - pips['p']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51b81e99",
   "metadata": {},
   "source": [
    "Then, switch to R kernel and define a function `roc_data` which takes PIP values from the PIP table of each method as inputs and add confidence level 5% as cutoff, then outputs ROC data.\n",
    "```R\n",
    "roc_data = function(dat, cols, cutoff = c(0.05, 0.999), connect_org = TRUE, grid = 100) {\n",
    "    d1 = dat[,cols]\n",
    "    if (connect_org) start = 0\n",
    "    else start = 1\n",
    "    ttv = (start:grid)/grid\n",
    "    ttv = ttv[which(ttv>=cutoff[1] & ttv<=cutoff[2])]\n",
    "    rst1 = t(sapply(ttv, function(x) c(sum(d1[,2][d1[,1]>=x]), length(d1[,2][d1[,1]>=x]), sum(d1[,2][d1[,1]>=x]==0))))\n",
    "    rst1 = cbind(rst1, sum(d1[,2]), sum(1-d1[,2]))\n",
    "    rst1 = as.data.frame(rst1)\n",
    "    colnames(rst1) = c('true_positive', 'total_positive', 'false_positive', 'total_signal', 'total_null')\n",
    "    rst2 = as.data.frame(cbind(rst1$true_positive / rst1$total_positive, rst1$true_positive / rst1$total_signal,  ttv))\n",
    "    rst3 = as.data.frame(cbind(1 - rst1$false_positive / rst1$total_null, rst1$true_positive / rst1$total_signal,  ttv))\n",
    "    if (connect_org) {\n",
    "        # make a stair to origin\n",
    "        rst2 = rbind(rst2, c(max(1 - cutoff[1], rst2[nrow(rst2),1]), max(rst2[nrow(rst2),2]-0.01, 0), rst2[nrow(rst2),3]))\n",
    "        rst2 = rbind(rst2, c(1, 0, 1))\n",
    "        rst3 = rbind(rst3, c(1, 0, 1))\n",
    "    }\n",
    "    colnames(rst2) = c('Precision', 'Recall', 'Threshold')\n",
    "    colnames(rst3) = c('TN', 'TP', 'Threshold')\n",
    "    return(list(counts = rst1, pr = rst2, roc = rst3))\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad53ff4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%get pips m1 m2 m3 m4 m5 m6\n",
    "res = list()\n",
    "for (m in c(m1,m2,m3,m4,m5,m6)) res[[m]] = roc_data(pips, c(paste0('pip_',m), 'is_signal'), cutoff = c(0, 1), connect_org = T, grid=200)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dbb4e22",
   "metadata": {},
   "source": [
    "Lastly, switch back to SoS kernel and `%get` ROC data from R kernel and plot them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65d7e04a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%get res --from R\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as font_manager\n",
    "font_prop = font_manager.FontProperties(size=24)\n",
    "\n",
    "fig, ax = plt.subplots(figsize = (10, 10))\n",
    "for m, c in zip([m1,m2,m3,m4,m5,m6], ['C0','C8','C3','C4','C2','C7']):\n",
    "    plt.plot(1 - res[m]['roc']['TN'], res[m]['roc']['TP'], c = c, label = m)\n",
    "plt.legend(loc = 'lower right', fontsize = 15)\n",
    "plt.ylabel(\"Power\", fontproperties=font_prop)\n",
    "plt.xlabel(\"False positive\", fontproperties=font_prop)\n",
    "plt.title('ROC curve', fontproperties=font_prop)\n",
    "plt.savefig(f\"{cwd}/ROC_curve.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "840ac601",
   "metadata": {},
   "source": [
    "$n=30,000$ |$n=80,000$ \n",
    ":---------:|:----------:\n",
    "![](cnv-figures/ROC_4simu_0802.png) | ![](cnv-figures/ROC_large_sample_0802_12.png)\n",
    "\n",
    "|![alt](cnv-figures/hybrid_11_simu123_0709.png) |![alt](cnv-figures/MCMC_11_simu123_0709.png)|\n",
    "|-|-|\n",
    "|![alt](cnv-figures/SuSiE_12_simu123_0709.png) | ![alt](cnv-figures/SIER_11_simu123_0709.png)"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "formats": "md:myst",
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": 0.13,
    "jupytext_version": "1.11.5"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "source_map": [
   14,
   214,
   222,
   230,
   249,
   253,
   281,
   285,
   288,
   304
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}