{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data transform: $y \\in \\{0,1\\}$ to $y \\in \\{0,t\\}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pprint import pprint\n",
    "from pandasql import sqldf\n",
    "from scipy import stats\n",
    "import pickle\n",
    "import feather\n",
    "from collections import Counter\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.32634787404 0.366383889501\n"
     ]
    }
   ],
   "source": [
    "# the lifetime prevalence of schizophrenia is 4.0/1000, not 1%, but we first set it as 0.01\n",
    "# according to https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1140960/\n",
    "prevalence = 0.01\n",
    "# relative risk is normally between 2 and 10\n",
    "RR = 2.5\n",
    "t = stats.norm(0,1).ppf(1-prevalence)\n",
    "# effect size of beta is z_mu, prior\n",
    "z_mu = t - stats.norm(0,1).ppf(1-RR*prevalence)\n",
    "print (t, z_mu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# create prior file\n",
    "cal_pthwy_genes = pd.read_table(\"../data/calciumgeneset.txt\", skiprows = 2, header = None, names = [\"gene_name\"])\n",
    "ref_genes = pd.read_table(\"data/refGene.txt.gz\", compression=\"gzip\", sep=\"\\t\", header = None, \n",
    "                          usecols=(2,12), names=[\"chrom\", \"gene_name\"]).drop_duplicates(subset=(\"gene_name\"))\n",
    "calpath_gene_pos = pd.merge(ref_genes, cal_pthwy_genes, how=\"inner\", on=[\"gene_name\"])\n",
    "diff = set(cal_pthwy_genes[\"gene_name\"].tolist()) - set(calpath_gene_pos[\"gene_name\"].tolist())\n",
    "# add the two missing genes and their chromosome in the dataframe\n",
    "calpath_gene_pos.loc[calpath_gene_pos.shape[0]] = [\"chr2\", list(diff)[0]]\n",
    "calpath_gene_pos.loc[calpath_gene_pos.shape[0]] = [\"chr15\", list(diff)[1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr.gene</th>\n",
       "      <th>prior.pr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17</td>\n",
       "      <td>CACNA1G</td>\n",
       "      <td>chr17.CACNA1G</td>\n",
       "      <td>0.266285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>PLCD4</td>\n",
       "      <td>chr2.PLCD4</td>\n",
       "      <td>0.107040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr20</td>\n",
       "      <td>PLCB4</td>\n",
       "      <td>chr20.PLCB4</td>\n",
       "      <td>0.362211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>CAMK2D</td>\n",
       "      <td>chr4.CAMK2D</td>\n",
       "      <td>0.140378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr14</td>\n",
       "      <td>SLC8A3</td>\n",
       "      <td>chr14.SLC8A3</td>\n",
       "      <td>0.169377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr8</td>\n",
       "      <td>PTK2B</td>\n",
       "      <td>chr8.PTK2B</td>\n",
       "      <td>0.210375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr9</td>\n",
       "      <td>GNA14</td>\n",
       "      <td>chr9.GNA14</td>\n",
       "      <td>0.059547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr8</td>\n",
       "      <td>VDAC3</td>\n",
       "      <td>chr8.VDAC3</td>\n",
       "      <td>0.157412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr12</td>\n",
       "      <td>PDE1B</td>\n",
       "      <td>chr12.PDE1B</td>\n",
       "      <td>0.075162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr12</td>\n",
       "      <td>CACNA1C</td>\n",
       "      <td>chr12.CACNA1C</td>\n",
       "      <td>0.256323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr3</td>\n",
       "      <td>ITPR1</td>\n",
       "      <td>chr3.ITPR1</td>\n",
       "      <td>0.271080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PRKACB</td>\n",
       "      <td>chr1.PRKACB</td>\n",
       "      <td>0.294886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr17</td>\n",
       "      <td>ATP2A3</td>\n",
       "      <td>chr17.ATP2A3</td>\n",
       "      <td>0.347966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr20</td>\n",
       "      <td>ADRA1D</td>\n",
       "      <td>chr20.ADRA1D</td>\n",
       "      <td>0.220837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr5</td>\n",
       "      <td>HTR4</td>\n",
       "      <td>chr5.HTR4</td>\n",
       "      <td>0.296342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PTAFR</td>\n",
       "      <td>chr1.PTAFR</td>\n",
       "      <td>0.304160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr4</td>\n",
       "      <td>CD38</td>\n",
       "      <td>chr4.CD38</td>\n",
       "      <td>0.004339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrX</td>\n",
       "      <td>SLC25A6</td>\n",
       "      <td>chrX.SLC25A6</td>\n",
       "      <td>0.333085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PTGER3</td>\n",
       "      <td>chr1.PTGER3</td>\n",
       "      <td>0.075069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr19</td>\n",
       "      <td>SPHK2</td>\n",
       "      <td>chr19.SPHK2</td>\n",
       "      <td>0.173236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr12</td>\n",
       "      <td>ATP2A2</td>\n",
       "      <td>chr12.ATP2A2</td>\n",
       "      <td>0.208458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr12</td>\n",
       "      <td>P2RX2</td>\n",
       "      <td>chr12.P2RX2</td>\n",
       "      <td>0.173613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CACNA1E</td>\n",
       "      <td>chr1.CACNA1E</td>\n",
       "      <td>0.323274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ATP2B4</td>\n",
       "      <td>chr1.ATP2B4</td>\n",
       "      <td>0.093543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr2</td>\n",
       "      <td>ADCY3</td>\n",
       "      <td>chr2.ADCY3</td>\n",
       "      <td>0.310170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr16</td>\n",
       "      <td>PLCG2</td>\n",
       "      <td>chr16.PLCG2</td>\n",
       "      <td>0.139039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr18</td>\n",
       "      <td>GNAL</td>\n",
       "      <td>chr18.GNAL</td>\n",
       "      <td>0.226588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ITPR3</td>\n",
       "      <td>chr6.ITPR3</td>\n",
       "      <td>0.100228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RYR2</td>\n",
       "      <td>chr1.RYR2</td>\n",
       "      <td>0.267196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr10</td>\n",
       "      <td>CAMK2G</td>\n",
       "      <td>chr10.CAMK2G</td>\n",
       "      <td>0.027177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>chr12</td>\n",
       "      <td>P2RX7</td>\n",
       "      <td>chr12.P2RX7</td>\n",
       "      <td>0.302376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>chr15</td>\n",
       "      <td>ITPKA</td>\n",
       "      <td>chr15.ITPKA</td>\n",
       "      <td>0.320564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>chr8</td>\n",
       "      <td>TRHR</td>\n",
       "      <td>chr8.TRHR</td>\n",
       "      <td>0.134721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CACNA1S</td>\n",
       "      <td>chr1.CACNA1S</td>\n",
       "      <td>0.259831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>chr9</td>\n",
       "      <td>GNAQ</td>\n",
       "      <td>chr9.GNAQ</td>\n",
       "      <td>0.275942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>chr15</td>\n",
       "      <td>PLCB2</td>\n",
       "      <td>chr15.PLCB2</td>\n",
       "      <td>0.121064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>chr7</td>\n",
       "      <td>ADCY1</td>\n",
       "      <td>chr7.ADCY1</td>\n",
       "      <td>0.343054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>chr15</td>\n",
       "      <td>RYR3</td>\n",
       "      <td>chr15.RYR3</td>\n",
       "      <td>0.265300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>chr7</td>\n",
       "      <td>PHKG1</td>\n",
       "      <td>chr7.PHKG1</td>\n",
       "      <td>0.157181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>chrX</td>\n",
       "      <td>CYSLTR1</td>\n",
       "      <td>chrX.CYSLTR1</td>\n",
       "      <td>0.068701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>chr5</td>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>chr5.PDGFRB</td>\n",
       "      <td>0.293815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>chr4</td>\n",
       "      <td>SLC25A4</td>\n",
       "      <td>chr4.SLC25A4</td>\n",
       "      <td>0.070235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>chr8</td>\n",
       "      <td>PPP3CC</td>\n",
       "      <td>chr8.PPP3CC</td>\n",
       "      <td>0.356010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>chr2</td>\n",
       "      <td>LHCGR</td>\n",
       "      <td>chr2.LHCGR</td>\n",
       "      <td>0.094462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ADRB3</td>\n",
       "      <td>chr8.ADRB3</td>\n",
       "      <td>0.330471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>chr3</td>\n",
       "      <td>OXTR</td>\n",
       "      <td>chr3.OXTR</td>\n",
       "      <td>0.327985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>chr10</td>\n",
       "      <td>CALML5</td>\n",
       "      <td>chr10.CALML5</td>\n",
       "      <td>0.363580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HTR6</td>\n",
       "      <td>chr1.HTR6</td>\n",
       "      <td>0.337619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>chr20</td>\n",
       "      <td>PLCG1</td>\n",
       "      <td>chr20.PLCG1</td>\n",
       "      <td>0.244072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>chr20</td>\n",
       "      <td>MYLK2</td>\n",
       "      <td>chr20.MYLK2</td>\n",
       "      <td>0.164047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>chr7</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>chr7.CHRM2</td>\n",
       "      <td>0.268262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>chr11</td>\n",
       "      <td>P2RX3</td>\n",
       "      <td>chr11.P2RX3</td>\n",
       "      <td>0.051683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>chrX</td>\n",
       "      <td>CACNA1F</td>\n",
       "      <td>chrX.CACNA1F</td>\n",
       "      <td>0.190121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>chrX</td>\n",
       "      <td>SLC25A5</td>\n",
       "      <td>chrX.SLC25A5</td>\n",
       "      <td>0.180459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ATP2B3</td>\n",
       "      <td>chrX.ATP2B3</td>\n",
       "      <td>0.180794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>chr16</td>\n",
       "      <td>PRKCB</td>\n",
       "      <td>chr16.PRKCB</td>\n",
       "      <td>0.272680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CALML6</td>\n",
       "      <td>chr1.CALML6</td>\n",
       "      <td>0.147582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>chr13</td>\n",
       "      <td>CYSLTR2</td>\n",
       "      <td>chr13.CYSLTR2</td>\n",
       "      <td>0.222805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>chr2</td>\n",
       "      <td>CHP</td>\n",
       "      <td>chr2.CHP</td>\n",
       "      <td>0.351839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>chr15</td>\n",
       "      <td>LOC729317</td>\n",
       "      <td>chr15.LOC729317</td>\n",
       "      <td>0.328145</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>178 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom  gene_name         chr.gene  prior.pr\n",
       "0    chr17    CACNA1G    chr17.CACNA1G  0.266285\n",
       "1     chr2      PLCD4       chr2.PLCD4  0.107040\n",
       "2    chr20      PLCB4      chr20.PLCB4  0.362211\n",
       "3     chr4     CAMK2D      chr4.CAMK2D  0.140378\n",
       "4    chr14     SLC8A3     chr14.SLC8A3  0.169377\n",
       "5     chr8      PTK2B       chr8.PTK2B  0.210375\n",
       "6     chr9      GNA14       chr9.GNA14  0.059547\n",
       "7     chr8      VDAC3       chr8.VDAC3  0.157412\n",
       "8    chr12      PDE1B      chr12.PDE1B  0.075162\n",
       "9    chr12    CACNA1C    chr12.CACNA1C  0.256323\n",
       "10    chr3      ITPR1       chr3.ITPR1  0.271080\n",
       "11    chr1     PRKACB      chr1.PRKACB  0.294886\n",
       "12   chr17     ATP2A3     chr17.ATP2A3  0.347966\n",
       "13   chr20     ADRA1D     chr20.ADRA1D  0.220837\n",
       "14    chr5       HTR4        chr5.HTR4  0.296342\n",
       "15    chr1      PTAFR       chr1.PTAFR  0.304160\n",
       "16    chr4       CD38        chr4.CD38  0.004339\n",
       "17    chrX    SLC25A6     chrX.SLC25A6  0.333085\n",
       "18    chr1     PTGER3      chr1.PTGER3  0.075069\n",
       "19   chr19      SPHK2      chr19.SPHK2  0.173236\n",
       "20   chr12     ATP2A2     chr12.ATP2A2  0.208458\n",
       "21   chr12      P2RX2      chr12.P2RX2  0.173613\n",
       "22    chr1    CACNA1E     chr1.CACNA1E  0.323274\n",
       "23    chr1     ATP2B4      chr1.ATP2B4  0.093543\n",
       "24    chr2      ADCY3       chr2.ADCY3  0.310170\n",
       "25   chr16      PLCG2      chr16.PLCG2  0.139039\n",
       "26   chr18       GNAL       chr18.GNAL  0.226588\n",
       "27    chr6      ITPR3       chr6.ITPR3  0.100228\n",
       "28    chr1       RYR2        chr1.RYR2  0.267196\n",
       "29   chr10     CAMK2G     chr10.CAMK2G  0.027177\n",
       "..     ...        ...              ...       ...\n",
       "148  chr12      P2RX7      chr12.P2RX7  0.302376\n",
       "149  chr15      ITPKA      chr15.ITPKA  0.320564\n",
       "150   chr8       TRHR        chr8.TRHR  0.134721\n",
       "151   chr1    CACNA1S     chr1.CACNA1S  0.259831\n",
       "152   chr9       GNAQ        chr9.GNAQ  0.275942\n",
       "153  chr15      PLCB2      chr15.PLCB2  0.121064\n",
       "154   chr7      ADCY1       chr7.ADCY1  0.343054\n",
       "155  chr15       RYR3       chr15.RYR3  0.265300\n",
       "156   chr7      PHKG1       chr7.PHKG1  0.157181\n",
       "157   chrX    CYSLTR1     chrX.CYSLTR1  0.068701\n",
       "158   chr5     PDGFRB      chr5.PDGFRB  0.293815\n",
       "159   chr4    SLC25A4     chr4.SLC25A4  0.070235\n",
       "160   chr8     PPP3CC      chr8.PPP3CC  0.356010\n",
       "161   chr2      LHCGR       chr2.LHCGR  0.094462\n",
       "162   chr8      ADRB3       chr8.ADRB3  0.330471\n",
       "163   chr3       OXTR        chr3.OXTR  0.327985\n",
       "164  chr10     CALML5     chr10.CALML5  0.363580\n",
       "165   chr1       HTR6        chr1.HTR6  0.337619\n",
       "166  chr20      PLCG1      chr20.PLCG1  0.244072\n",
       "167  chr20      MYLK2      chr20.MYLK2  0.164047\n",
       "168   chr7      CHRM2       chr7.CHRM2  0.268262\n",
       "169  chr11      P2RX3      chr11.P2RX3  0.051683\n",
       "170   chrX    CACNA1F     chrX.CACNA1F  0.190121\n",
       "171   chrX    SLC25A5     chrX.SLC25A5  0.180459\n",
       "172   chrX     ATP2B3      chrX.ATP2B3  0.180794\n",
       "173  chr16      PRKCB      chr16.PRKCB  0.272680\n",
       "174   chr1     CALML6      chr1.CALML6  0.147582\n",
       "175  chr13    CYSLTR2    chr13.CYSLTR2  0.222805\n",
       "176   chr2        CHP         chr2.CHP  0.351839\n",
       "177  chr15  LOC729317  chr15.LOC729317  0.328145\n",
       "\n",
       "[178 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "calpath_gene_pos[\"chr.gene\"] = calpath_gene_pos.apply(lambda row: \"{}.{}\".format(row[\"chrom\"], row[\"gene_name\"]), \n",
    "                                                      axis=1)\n",
    "calpath_gene_pos[\"prior.pr\"] = calpath_gene_pos.apply(lambda row: np.random.uniform(0, z_mu), axis=1)\n",
    "calpath_gene_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "beta = z_mu\n",
    "y_tran = t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "filename = \"data/del_sim_scale3_shape1_N100_M100_999.feather\"\n",
    "fileout = \"data/del_sim_scale3_shape1_N100_M100_999.dap\"\n",
    "\n",
    "input_file = feather.read_dataframe(filename)\n",
    "input_file[\"phenotype\"] = input_file.apply(lambda row: y_tran if row[\"phenotype\"]==1.0 else 0, axis=1)\n",
    "\n",
    "def run_dap(df, fileout, multiplier = 2, prefix = None, exec_path = None):\n",
    "    '''Convert pandas dataframe to dap input:\n",
    "        - phenotype / genotype file\n",
    "        - prior file\n",
    "        - grid file (of effect size): omega^2 + phi^2 is what we care. Let's set it to\n",
    "            *************** for now, as we only have one Y\n",
    "    '''\n",
    "    print (str(datetime.now()))\n",
    "    import os\n",
    "    if prefix is None:\n",
    "        import time\n",
    "        prefix = \"/tmp/F\" + str(time.time())\n",
    "    if exec_path is None:\n",
    "        exec_path = 'dap/dap'\n",
    "#     chrom = [\"chr{}\".format(i) for i in list(range(1,23))+[\"X\"]]\n",
    "#     pos is all the gene names showed in .feather\n",
    "    pos = df.columns.values[1:].tolist()\n",
    "#     first item in \"dat\" is phenotype for all samples (# of rows in .feather)\n",
    "    dat = [['pheno', 'trait', 'chicago'] + [str(x) for x in df['phenotype']]]\n",
    "    prior = []\n",
    "    grid = [(0,beta)]\n",
    "#     item is gene name\n",
    "    for idx, item in enumerate(df.columns.values):\n",
    "        if item == \"phenotype\":\n",
    "            continue\n",
    "        dat.append(['geno', '{}.{}'.format(ref_genes[ref_genes[\"gene_name\"]==item][\"chrom\"].tolist()[0], item), \n",
    "                    'chicago'] + [str(x) for x in df[item]])\n",
    "        n_overlap_gene = len(set(cal_pthwy_genes[\"gene_name\"]) & set(pos))\n",
    "        prior_causal = multiplier / len(set(pos))\n",
    "        prior_noncausal = (1-prior_causal*n_overlap_gene) / (len(set(pos))-n_overlap_gene)\n",
    "        prior_pr = str(prior_causal) if item in cal_pthwy_genes[\"gene_name\"].tolist() else str(prior_noncausal)\n",
    "        prior.append(['{}.{}'.format(ref_genes[ref_genes[\"gene_name\"]==item][\"chrom\"].tolist()[0], item), prior_pr])\n",
    "#     print (prior_causal, prior_noncausal)\n",
    "    with open(prefix + '.dat', 'w') as f:\n",
    "        f.write('\\n'.join([' '.join(x) for x in dat]))\n",
    "    with open(prefix + '.prior', 'w') as f:\n",
    "        f.write('\\n'.join([' '.join(x) for x in prior]))\n",
    "    with open(prefix + '.grid', 'w') as f:\n",
    "        f.write('\\n'.join([' '.join(map(str, x)) for x in grid]))\n",
    "    os.system(\"{0} -d {1}.dat -g {1}.grid -t 8 -it 0.05 -prior {1}.prior > {2}\".format(exec_path, prefix, fileout))\n",
    "    print (str(datetime.now()))\n",
    "#     return prior"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2017-08-04 20:22:54.007086\n",
      "2017-08-04 20:23:02.148236\n"
     ]
    }
   ],
   "source": [
    "testp=run_dap(input_file, fileout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "calpath_gene_pos.to_csv(\"test.csv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prior_c = len(set(cal_pthwy_genes[\"gene_name\"]) & set(df.columns.values[1:].tolist())) *2 / len(set(df.columns.values[1:].tolist())) / len(set(cal_pthwy_genes[\"gene_name\"]))\n",
    "prior_c = 2/len(set(df.columns.values[1:].tolist()))\n",
    "n_overlap_gene = len(set(cal_pthwy_genes[\"gene_name\"]) & set(df.columns.values[1:].tolist()))\n",
    "prior_nc = (1-prior_c*n_overlap_gene)/(len(set(df.columns.values[1:].tolist()))-n_overlap_gene)\n",
    "for idx, item in enumerate(df.columns.values):\n",
    "    prior_t = str(prior_c) if item in cal_pthwy_genes[\"gene_name\"].tolist() else str(prior_nc)\n",
    "#     print (prior_t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX']\n"
     ]
    }
   ],
   "source": [
    "# chrom\n",
    "chrom = [\"chr{}\".format(i) for i in list(range(1,23))+[\"X\"]]\n",
    "print ((chrom))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
